File: backend/src/main/java/com/andrew/hdss/models/Individual.java
package com.andrew.hdss.models;

import com.andrew.hdss.models.enums.Sex;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;

@Getter
@Setter
@Entity
@Table(name = "individuals")
public class Individual extends SyncableEntity {

    @Column(nullable = false, unique = true)
    private String extendedId;

    @Column(nullable = false)
    private String firstName;

    @Column(nullable = false)
    private String lastName;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Sex sex;

    @Column(nullable = false)
    private LocalDate dateOfBirth;

    @Column(nullable = false)
    private boolean dobEstimated = false;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "mother_id")
    private Individual mother;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "father_id")
    private Individual father;
}

File: android/app/src/main/java/com/andrew/hdss/data/models/Individual.kt
package com.andrew.hdss.data.models

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import com.andrew.hdss.data.models.enums.Sex
import java.time.LocalDate

@Entity(
    tableName = "individuals",
    foreignKeys = [
        ForeignKey(
            entity = Individual::class,
            parentColumns = ["clientId"],
            childColumns = ["motherClientId"],
            onDelete = ForeignKey.SET_NULL
        ),
        ForeignKey(
            entity = Individual::class,
            parentColumns = ["clientId"],
            childColumns = ["fatherClientId"],
            onDelete = ForeignKey.SET_NULL
        )
    ],
    indices = [Index("motherClientId"), Index("fatherClientId"), Index("serverId")]
)
data class Individual(
    @PrimaryKey
    val clientId: String,
    val serverId: Long?,
    val extendedId: String?,
    val firstName: String,
    val lastName: String,
    val sex: Sex,
    val dateOfBirth: LocalDate?,
    val dobEstimated: Boolean,
    val motherClientId: String?,
    val fatherClientId: String?
)